server/src/test/java/org/elasticsearch/cluster/coordination/ClusterBootstrapServiceTests.java

```diff
@@ -735,6 +735,27 @@ public void testBootstrapStateLogging() {
             }
 
             mockAppender.assertAllExpectationsMatched();
+
+            mockAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "bootstrapped node message if discovery type is single node ",
+                    ClusterBootstrapService.class.getCanonicalName(),
+                    Level.INFO,
+                    "this node is locked into single-node cluster UUID [test-uuid]"
+                )
+            );
+
+            new ClusterBootstrapService(
+                Settings.builder().put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), DiscoveryModule.SINGLE_NODE_DISCOVERY_TYPE).build(),
+                transportService,
+                Collections::emptyList,
+                () -> false,
+                vc -> {
+                    throw new AssertionError("should not be called");
+                }
+            ).logBootstrapState(Metadata.builder().clusterUUID("test-uuid").clusterUUIDCommitted(true).build());
+
+            mockAppender.assertAllExpectationsMatched();
```

Comment:
This does not verify that the bug is fixed, i.e. that the message is only emitted once. It would be sufficient to verify that `deterministicTaskQueue.runAllTasks()` completes. You'll need to do that in a separate test case, or at least earlier in this one, because there's no way to shut down the repeated logging from the badly-configured `ClusterBootstrapService` instance.